File: Backend/Backend/Services/CardService.cs

using Backend.Models.ViewModels;
using Backend.Repositories;

namespace Backend.Services;

public interface ICardService
{
    Task<List<CardViewModel>> GetCards(string? searchTerm);
    Task<CardViewModel> AddCard(CardViewModel card);
}

public class CardService(ICardRepository cardRepository) : ICardService
{
    private readonly ICardRepository _cardRepository = cardRepository;

    public async Task<CardViewModel> AddCard(CardViewModel card)
    {
        return await _cardRepository.AddCard(card);
    }

    public async Task<List<CardViewModel>> GetCards(string? searchTerm)
    {
        return await _cardRepository.GetCards(searchTerm);
    }
}

File: Frontend/src/app/card/services/card.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import { HttpClient } from '@angular/common/http';
import { Card } from '../models/card.model';
import { map } from 'rxjs';

@Injectable()
export class CardService {
  private readonly _baseApiUrl: string;

  constructor(private readonly _httpClient: HttpClient) {
    this._baseApiUrl = environment.apiBaseUrl;
  }

  getCards(searchTerm?: string) {
    return this._httpClient
      .get<Card[]>(
        `${this._baseApiUrl}/api/Card/GetCards${
          searchTerm ? `?searchTerm=${searchTerm}` : ''
        }`
      )
      .pipe(map((cards) => cards.map((card) => new Card(card))));
  }

  addCard(newcard: Card) {
    return this._httpClient
      .post<Card>(`${this._baseApiUrl}/api/Card/AddCard`, newcard)
      .pipe(map((card) => new Card(card)));
  }
}
